precompiles/pointer/pointer.go

```diff
@@ -0,0 +1,257 @@
+	"github.com/sei-protocol/sei-chain/x/evm/artifacts/cw20"
+	"github.com/sei-protocol/sei-chain/x/evm/artifacts/cw721"
+	"github.com/sei-protocol/sei-chain/x/evm/artifacts/native"
+)
+
+const (
+	AddNativePointer = "addNativePointer"
+	AddCW20Pointer   = "addCW20Pointer"
+	AddCW721Pointer  = "addCW721Pointer"
+)
+
+const PointerAddress = "0x000000000000000000000000000000000000100b"
+
+var _ vm.PrecompiledContract = &Precompile{}
+
+// Embed abi json file to the executable binary. Needed when importing as dependency.
+//
+//go:embed abi.json
+var f embed.FS
+
+type Precompile struct {
+	pcommon.Precompile
+	evmKeeper   pcommon.EVMKeeper
+	bankKeeper  pcommon.BankKeeper
+	wasmdKeeper pcommon.WasmdViewKeeper
+	address     common.Address
+
+	AddNativePointerID []byte
+	AddCW20PointerID   []byte
+	AddCW721PointerID  []byte
+}
+
+func NewPrecompile(evmKeeper pcommon.EVMKeeper, bankKeeper pcommon.BankKeeper, wasmdKeeper pcommon.WasmdViewKeeper) (*Precompile, error) {
+	abiBz, err := f.ReadFile("abi.json")
+	if err != nil {
+		return nil, fmt.Errorf("error loading the pointer ABI %s", err)
+	}
+
+	newAbi, err := ethabi.JSON(bytes.NewReader(abiBz))
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Precompile{
+		Precompile:  pcommon.Precompile{ABI: newAbi},
+		evmKeeper:   evmKeeper,
+		bankKeeper:  bankKeeper,
+		wasmdKeeper: wasmdKeeper,
+		address:     common.HexToAddress(PointerAddress),
+	}
+
+	for name, m := range newAbi.Methods {
+		switch name {
+		case AddNativePointer:
+			p.AddNativePointerID = m.ID
+		case AddCW20Pointer:
+			p.AddCW20PointerID = m.ID
+		case AddCW721Pointer:
+			p.AddCW721PointerID = m.ID
+		}
+	}
+
+	return p, nil
+}
+
+// RequiredGas returns the required bare minimum gas to execute the precompile.
+func (p Precompile) RequiredGas(input []byte) uint64 {
+	// gas is calculated dynamically
+	return 0
+}
+
+func (p Precompile) Address() common.Address {
+	return p.address
+}
+
+func (p Precompile) RunAndCalculateGas(evm *vm.EVM, caller common.Address, _ common.Address, input []byte, suppliedGas uint64, value *big.Int) (ret []byte, remainingGas uint64, err error) {
+	ctx, method, args, err := p.Prepare(evm, input)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	switch method.Name {
+	case AddNativePointer:
+		return p.AddNative(ctx, method, caller, args, value, evm, suppliedGas)
+	case AddCW20Pointer:
+		return p.AddCW20(ctx, method, caller, args, value, evm, suppliedGas)
+	case AddCW721Pointer:
+		return p.AddCW721(ctx, method, caller, args, value, evm, suppliedGas)
+	}
+	return
+}
+
+func (p Precompile) Run(*vm.EVM, common.Address, []byte, *big.Int) ([]byte, error) {
+	panic("static gas Run is not implemented for dynamic gas precompile")
+}
+
+func (p Precompile) AddNative(ctx sdk.Context, method *ethabi.Method, caller common.Address, args []interface{}, value *big.Int, evm *vm.EVM, suppliedGas uint64) (ret []byte, remainingGas uint64, err error) {
+	pcommon.AssertArgsLength(args, 1)
+	token := args[0].(string)
```

Comment:
what happens if they pass no args?

Reply:
the call would panic and fail (the panic would be caught on the DeliverTx level)